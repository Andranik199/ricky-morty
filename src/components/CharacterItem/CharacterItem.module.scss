.card {
  max-width: 250px;
  flex: 1 0 21%;
  margin: 5px
}
