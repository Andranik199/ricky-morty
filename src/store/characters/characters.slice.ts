import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export enum Gender {
    MALE = 'Male',
    FEMALE = 'Female'
}

export enum Status {
    ALIVE = 'Alive',
    DEAD = 'Dead'
}

type Episode = {
    id: string;
    name: string;
    air_date: string;
}
export type Character = {
    id: string;
    name: string;
    status: Status;
    species: string;
    type: string;
    gender: Gender;
    image: string;
    location: { name: string },
    episode?: Array<Episode>
}
type InitialState = {
    items: Array<Character>;
    totalPages: number;
    isPending: boolean;
    specificCharacter: Character | null;
}
const initialState: InitialState = {
  items: [],
  totalPages: 0,
  isPending: false,
  specificCharacter: null,
};
export const charactersSlice = createSlice({
  name: 'characters',
  initialState,
  reducers: {
    setSpecificCharacter: (state, action: PayloadAction<Character>) => {
      return {
        ...state,
        specificCharacter: action.payload,
      };
    },
    setCharactersPendingStatus: (state, action) => {
      return {
        ...state,
        isPending: action.payload,
      };
    },
    setCharacters: (state, action) => {
      return {
        ...state,
        items: action.payload,
      };
    },
    setTotalPages: (state, action) => {
      return {
        ...state,
        totalPages: action.payload,
      };
    },
  },
});
