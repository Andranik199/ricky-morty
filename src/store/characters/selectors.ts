import { createSelector } from '@reduxjs/toolkit';
import { RootState } from '../index';


const CharacterSelector = (state: RootState) => state.characters;

export const selectCharacters = createSelector(CharacterSelector, (state) => state.items || []);
export const selectPagesCount = createSelector(CharacterSelector, (state) => state.totalPages);
export const selectCharactersPendingStatus = createSelector(CharacterSelector, (state) => state.isPending);
export const selectSpecificCharacter = createSelector(CharacterSelector, (state) => state.specificCharacter);
