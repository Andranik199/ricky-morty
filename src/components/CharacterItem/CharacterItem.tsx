import React from 'react';
import { Card, CardActionArea, CardContent, CardHeader, CardMedia, Typography } from '@mui/material';
import { Character, Status } from 'store/characters/characters.slice';
import { GiDeathSkull } from 'react-icons/gi';

type CharacterItemProps = {
    item: Character;
    handleSelect: (id: string) => void;
}
export const CharacterItem: React.FC<CharacterItemProps> = ({ item, handleSelect }) => {
  const handleClick = () => {
    handleSelect?.(item.id);
  };
  return (
    <div data-testid='characterItem' onClick={handleClick}>
      <Card
        key={item.id}
        sx={{
          maxWidth: 250,
          flex: '1 0 21%',
          margin: '5px',
        }}>
        <CardHeader
          avatar={
            item.status === Status.DEAD && <GiDeathSkull/>
          }
          title={`Gender - ${item.gender}`}
          subheader={`Location ${item.location.name}`}
        />
        <CardActionArea>
          <CardMedia
            component="img"
            height="140"
            image={item.image}
            alt="green iguana"
          />
          <CardContent>
            <Typography gutterBottom variant="h5" component="div">
              {item.name}
            </Typography>
          </CardContent>
        </CardActionArea>
      </Card>
    </div>
  );
};
