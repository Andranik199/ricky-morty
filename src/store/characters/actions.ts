import { generateGetCharactersQuery, generateGetSpecificCharacterQuery } from 'common/query';
import { charactersActions } from './config';
import { Query, RootState } from '../index';
import { ThunkAction } from '@reduxjs/toolkit';
import { Character } from './characters.slice';

type CharactersResponse = {
    results: Array<Character>;
    info: {
        pages: number;
    }
}
type Thunk = ThunkAction<void, RootState, { query: Query }, any>
export type GetCharactersFilters = Partial<{
    status: string;
    name: string;
}>
export const getCharacters = (page = 1, filters?: GetCharactersFilters): Thunk => (dispatch, getState, { query }) => {
  dispatch(charactersActions.setCharactersPendingStatus(true));
  query<unknown, CharactersResponse>('characters', generateGetCharactersQuery(page, filters))
      .then((response) => {
        dispatch(charactersActions.setCharacters(response.results));
        dispatch(charactersActions.setTotalPages(response.info.pages));
      }).finally(() => dispatch(charactersActions.setCharactersPendingStatus(false)));
};


export const getCharacterById = (id: string): Thunk => (dispatch, getState, { query }) => {
  query<unknown, Character>('character', generateGetSpecificCharacterQuery(id)).then((res) => {
    dispatch(charactersActions.setSpecificCharacter(res));
  });
};
