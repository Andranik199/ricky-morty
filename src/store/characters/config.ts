import { charactersSlice } from './characters.slice';
import * as actions from './actions';


export const charactersActions = {
  ...charactersSlice.actions,
  ...actions,
};


export * as charactersSelectors from './selectors';
