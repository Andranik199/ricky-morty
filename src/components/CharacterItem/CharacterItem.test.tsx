import { create } from 'react-test-renderer';
import { CharacterItem } from './CharacterItem';
import { Gender, Status } from 'store/characters/characters.slice';

it('CharacterItem component should render properly', () => {
  const component = create(<CharacterItem item={{
    id: '1',
    name: 'Rick Sanchez',
    status: Status.ALIVE,
    species: 'Human',
    type: '',
    gender: Gender.MALE,
    location: {
      name: 'Test location',
    },
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
  }} handleSelect={jest.fn()}/>);
  expect(component.toJSON()).toMatchSnapshot();
});
