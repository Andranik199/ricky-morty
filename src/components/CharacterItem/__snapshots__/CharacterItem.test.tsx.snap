// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterItem component should render properly 1`] = `
<div
  data-testid="characterItem"
  onClick={[Function]}
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-qlmfal-MuiPaper-root-MuiCard-root"
  >
    <div
      className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          className="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
        >
          Gender - Male
        </span>
        <span
          className="MuiTypography-root MuiTypography-body1 MuiCardHeader-subheader css-nrdprl-MuiTypography-root"
        >
          Location Test location
        </span>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <img
        alt="green iguana"
        className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="140"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
      <div
        className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Rick Sanchez
        </div>
      </div>
      <span
        className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
      />
    </button>
  </div>
</div>
`;
